blurring_as_a_service/cleaning_pipeline/source/sampling_pipeline.py

```diff
@@ -0,0 +1,458 @@
+        bin_size = SmartSampling.determine_bin_size(detection_range)
+
+        # Calculate the bin edges
+        bins = np.linspace(min_count, max_count, bin_size + 1)
+
+        # Initialize a dictionary to hold bin counts and labels
+        bin_counts = {}
+        
+        # Iterate through the given bins array and create a list of bin labels. 
+        # Each label represents a range, formatted as "start-end", 
+        # where "start" is the beginning of a bin and "end" is one less than the start of the next bin.
+        bin_labels = [f"{int(bins[i])}-{int(bins[i + 1]) - 1}" for i in range(len(bins) - 1)]
+
+        # Categorize images into bins
+        unique_df['bin_label'] = pd.cut(unique_df['count'], bins, labels=bin_labels, include_lowest=True, right=True)
+
+        for label in bin_labels:
+            bin_counts[label] = unique_df[unique_df['bin_label'] == label]
+
+        return unique_df, bin_counts, bin_labels
+
+    @staticmethod
+    def sample_images_equally_from_bins(
+        df: pd.DataFrame, 
+        percentage_ratio: float
+    ) -> Dict[datetime, List[Tuple[str, datetime, str]]]:
+        """
+        Samples a percentage of images equally from each bin for each date.
+
+        Parameters
+        ----------
+        df : pd.DataFrame
+            A DataFrame containing image data with columns including 'image_upload_date', 'image_customer_name', 'image_filename', 'bin_label'.
+        percentage_ratio : float
+            The ratio of total images to sample from each date.
+
+        Returns
+        -------
+        pd.DataFrame
+            A DataFrame of sampled images.
+        """
+        
+        # Initialize the DataFrame to store sampled images
+        sampled_images_df = pd.DataFrame()
+
+        # Extract all unique upload dates from the DataFrame
+        unique_dates = df['image_upload_date'].unique()
+        logger.info(f'Unique dates: {unique_dates}')
+
+        for upload_date in unique_dates:
+            # Filter for the current date
+            df_date = df[df['image_upload_date'] == upload_date]
+            total_images = len(df_date)
+            logger.info(f'Total images on date {upload_date}: {total_images}')
+            
+            # Calculate the total number of images to sample
+            total_images_to_sample = int(total_images * percentage_ratio)
+            
+            # Ensure at least one image is sampled if total_images_to_sample > 0
+            total_images_to_sample = max(total_images_to_sample, 1) if total_images > 0 else 0
+            
+            logger.info(f'Total images to sample on date {upload_date}: {total_images_to_sample}')
+            
+            # Check if 'bin_label' column exists
+            if 'bin_label' not in df_date.columns:
+                logger.error("bin_label column not found in DataFrame for date: " + str(upload_date))
+                continue  # Skip to next date or handle error as needed
```

Comment:
Not a big fan of continues and breaks, there is a lot of discussion if they are bad practice or not. 
I found easier to follow the flow of the program avoiding them.
@epureanudiana what is your idea about it?

Reply:
yea, previously we replaced it with if else